### PR TITLE
Refactor STAB (again)

### PR DESCRIPTION
**Pull request by Karthik99999**
Originally was just trying to fix a bug with the pledge moves and tera stellar (combined pledge moves should always be boosted to 2x by tera stellar), but may have gotten a little carried away. Anyway, this changes the STAB code to much more closely resemble the actual STAB code from the games, and introduces a new `ModifySTAB` event to further improve the separation and readability of the Adaptability logic.

### CODE REVIEW

File: data/conditions.ts [hunk 1]
```diff
@@ -388,17 +388,7 @@ export const Conditions: {[k: string]: ConditionData} = {
 			target.removeVolatile('Protect');
 			target.removeVolatile('Endure');
 
-			if (data.source.hasAbility('infiltrator') && this.gen >= 6) {
-				data.moveData.infiltrates = true;
-			}
-			if (data.source.hasAbility('normalize') && this.gen >= 6) {
-				data.moveData.type = 'Normal';
-			}
```
Comment (urkerab) [inline data/conditions.ts:396]: Why this change?

Comment (Karthik99999) [reply data/conditions.ts]: I removed them all thinking they were all redundant, but apparently the `onModifyMove` and `onModifyType` events don't run which is why they were there in the first place (and why Adaptability doesn't need to be there anymore since that's on a new event). I added them back now.

**Comment by KrisXV** [comment]
@Marty-D @DaWoblefet any opinions/input?

**Review by DaWoblefet — APPROVED** [review]

**Comment by DaWoblefet** [comment]
Thanks Karthik!
